std/assembly/string.ts

```diff
@@ -393,7 +394,143 @@ export class String {
     return result;
   }
 
+  /* TODO
+   * 1. regex for search
+   * 2. callback function for replacement
+   */
+  replace(search: String, replacement: String): String {
+    assert(this !== null && search !== null && replacement !== null);
+    var len = this.length;
+    var slen = search.length;
+    if (ASC_SHRINK_LEVEL < 1) {
+      if (len <= slen) {
+        if (len < slen) return this;
+        return select<String>(replacement, this, search == this);
+      }
+    }
+    var index = this.indexOf(search);
+    if (~index) {
+      if (ASC_SHRINK_LEVEL >= 1) {
+        return this.substring(0, index)
+          .concat(replacement)
+          .concat(this.substring(index + slen));
+      } else {
+        let rlen = replacement.length;
+        len = len - slen;
+        let olen = len + rlen;
+        if (olen) {
+          let result = allocateUnsafe(olen);
+          copyUnsafe(result, 0, this, 0, index);
+          copyUnsafe(result, index, replacement, 0, rlen);
+          copyUnsafe(result, rlen + index, this, index + slen, len - index);
+          return result;
+        }
+      }
+    }
+    return this;
+  }
+
+  /* TODO
+   * 1. regex for search
+   * 2. callback function for replacement
+   */
+  replaceAll(search: String, replacement: String): String {
+    assert(this !== null && search !== null && replacement !== null);
+    var len = this.length;
+    var slen = search.length;
+    if (ASC_SHRINK_LEVEL < 1) {
+      if (len <= slen) {
+        if (len < slen) return this;
+        return select<String>(replacement, this, search == this);
+      }
+    }
+    var rlen = replacement.length;
+    if (!slen) {
+      // Handle special case when we should insert replacement between each char
+      // TODO handle surrogate pairs as well
+      if (!rlen) return this;
+      let result = allocateUnsafe(len + (len + 1) * rlen);
+      copyUnsafe(result, 0, replacement, 0, rlen);
+      let offset = rlen;
+      for (let i = 0; i < len; ++i) {
+        store<u16>(
+          changetype<usize>(result) + (offset++ << 1),
+          load<u16>(changetype<usize>(this) + (i << 1), HEADER_SIZE),
+          HEADER_SIZE
+        );
+        copyUnsafe(result, offset, replacement, 0, rlen);
+        offset += rlen;
+      }
+      return result;
+    }
+    var prev = 0, next = 0;
+    if (ASC_SHRINK_LEVEL >= 1) {
```

Comment:
Similar. I think we should design these functions to be standalone.

Reply:
Done!